src/scf/self_consistent_field.jl

```diff
@@ -133,25 +93,152 @@ Solve the Kohn-Sham equations with a SCF algorithm, starting at ρ.
     end
 
     # Tolerance and maxiter are only dummy here: Convergence is flagged by is_converged
-    # inside the fixpoint_map. Also we do not use the return value of fpres but rather the
-    # one that got updated by fixpoint_map
-    fpres = solver(fixpoint_map, ρout, maxiter; tol=eps(T))
+    # inside the fixpoint_map.
+    solver(fixpoint_map, ρout, maxiter; tol=eps(T))
 
-    # We do not use the return value of fpres but rather the one that got updated by fixpoint_map
-    # ψ is consistent with ρout, so we return that. We also perform
-    # a last energy computation to return a correct variational energy
-    energies, ham = energy_hamiltonian(basis, ψ, occupation;
-                                       ρ=ρout, eigenvalues=eigenvalues, εF=εF)
+    # We do not use the return value of solver but rather the one that got updated by fixpoint_map
+    # ψ is consistent with ρout, so we return that. We also perform a last energy computation
+    # to return a correct variational energy
+    energies, ham = energy_hamiltonian(basis, ψ, occupation; ρ=ρout, eigenvalues, εF)
 
     # Measure for the accuracy of the SCF
     # TODO probably should be tracked all the way ...
     norm_Δρ = norm(info.ρout - info.ρin) * sqrt(basis.dvol)
 
     # Callback is run one last time with final state to allow callback to clean up
-    info = (; ham, basis, energies, converged, occupation_threshold,
-            ρ=ρout, α=damping, eigenvalues, occupation, εF,
-            n_iter, n_ep_extra, ψ, info.diagonalization,
-            stage=:finalize, algorithm="SCF", norm_Δρ)
+    info = (; ham, basis, energies, converged, bands.occupation_threshold,
+            ρ=ρout, α=damping, eigenvalues, occupation, εF, info.n_bands,
+            n_iter, ψ, info.diagonalization, stage=:finalize,
+            algorithm="SCF", norm_Δρ)
     callback(info)
     info
 end
+
+
+"""
+Obtain new density ρ by diagonalizing `ham`. Follows the policy imposed by the `bands`
+data structure to determine and adjust the number of bands to be computed.
+"""
+function next_density(ham::Hamiltonian, bands=AdaptiveBands(ham.basis.model);
+                      eigensolver=lobpcg_hyper, ψ=nothing, eigenvalues=nothing,
+                      occupation=nothing, kwargs...)
+    n_bands, n_bands_compute = determine_n_bands(bands, occupation, eigenvalues, ψ)
```

Comment:
it's a bit confusing to have n_bands, n_bands_compute... Would n_bands_converge be good? Or would it be too long? We don't use it that much...